Add additional tests for server dynamic resources

latha-subramanian opened the pull request:
<!--- Please review your changes in preview mode -->
<!--- Provide a general summary of your changes in the Title above -->

#### Bug/feature Description
Add automated tests to verify server dynamic resources

#### Affected Platform(s)
18.2.0

#### Cause / Analysis / Design
Missing automated tests for server dynamic resources verification

#### Solution Description
Added tests for following scenarios:
1.  Invalid values returned from server_dyn_resource script for resource type ‘size’.
2. Invalid values returned from server_dyn_resource script for resource type ‘string’. (2 tests)
3. Invalid values returned from server_dyn_resource script for resource type ‘string_array’.(2 tests)
4. Invalid values returned from server_dyn_resource script for resource type ‘float’.
5. Invalid values returned from server_dyn_resource script for resource type ‘boolean’.

#### Checklist:
<!--- Use the preview button to see the checkboxes/links properly. -->
<!--- Go over the following points, and put an `x` (without spaces around it) in the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] I have joined the **[pbspro community forum](http://community.pbspro.org/)**.
- [x] My pull request contains a **single, signed** commit. See **[setting up gpg signature](https://pbspro.atlassian.net/wiki/display/DG/Signing+Your+Git+Commits)**.
- [x] My code follows the **[coding style](https://pbspro.atlassian.net/wiki/display/DG/Coding+Standards)** of this project.
- [ ] My change requires project documentation. See **[required documentation checklist](https://pbspro.atlassian.net/wiki/display/DG/Checklist+for+Developing+Features+and+Bug+Fixes)** for details.
   - [ ] I have added documentation in the **[project documentation area](https://pbspro.atlassian.net/wiki/display/PD)**.
- [ ] I have added new **PTL test(s) to my commit**. (See **[using PTL for testing](https://pbspro.atlassian.net/wiki/display/DG/Using+PTL+for+Testing)**) *(or)*
   - [ ] I have added  **manual test(s) to this pull request and explained why PTL is not appropriate** for this case.
- [ ] All new and existing automated tests have passed. (See **[running automated PTL tests](https://pbspro.atlassian.net/wiki/display/DG/PTL+Quick+Start+Guide)**).
- [x] I have attached **test logs to this pull request** as evidence of testing/verification.

[server_dyn_res.log](https://github.com/PBSPro/pbspro/files/1921970/server_dyn_res.log)

__***For further information please visit the [Developer Guide Home](https://pbspro.atlassian.net/wiki/display/DG/Developer+Guide+Home).***__

dinesh1306 commented on test/tests/functional/pbs_svr_dyn_res.py:
```diff
@@ -480,3 +480,237 @@ def test_res_size_runtime(self):
+        # The job shouldn't run
+        job_comment = "Can Never Run: Insufficient amount of server resource:"
+        job_comment += " foo (a != 100)"
+        a = {'job_state': 'Q', 'comment': job_comment}
```
here a != 100 check on value base not on type . job is in Q state because of a != 100 , but not due to a is string and 100 is integer. Might you need to update test case.

latha-subramanian replied:
Removing this test as there is no value.

dinesh1306 commented on test/tests/functional/pbs_svr_dyn_res.py:
```diff
@@ -480,3 +480,237 @@ def test_res_size_runtime(self):
+
+        # The job shouldn't run
+        job_comment = "Can Never Run: Insufficient amount of server resource:"
+        job_comment += " foo (abc != 1.25)"
```
same as above.

bhroam commented on test/tests/functional/pbs_svr_dyn_res.py:
```diff
@@ -480,3 +480,237 @@ def test_res_size_runtime(self):
+        # Create a resource of type size
+        resname = ["foobar"]
+        restype = ["size"]
+        resflag = ["q"]
```
server_dyn_res resources are not supposed to have flag=q.  This will cause the resource to be double counted.  It doesn't really matter for this test, but we might as well set things up right.

latha-subramanian replied:
Removed flag

bhroam commented on test/tests/functional/pbs_svr_dyn_res.py:
```diff
@@ -480,3 +480,237 @@ def test_res_size_runtime(self):
+        job_comment = "Can Never Run: Insufficient amount of server resource:"
+        job_comment += " foobar (R: 2gb A: 0kb T: 0kb)"
+        a = {'job_state': 'Q', 'comment': job_comment}
+        self.server.expect(JOB, a, id=jid, attrop=PTL_AND)
```
You're probably doing the right test here, but I'll offer something else to look for as well.  There is an error log message printed at query time when a server_dyn_res script returns an invalid value ("Script <script> returned bad output").

latha-subramanian replied:
Added this check

bhroam commented on test/tests/functional/pbs_svr_dyn_res.py:
```diff
@@ -480,3 +480,237 @@ def test_res_size_runtime(self):
+        """
+        Test invalid values returned from server_dyn_resource
+        script for resource type 'string'
+        Script returns a 'integer' instead of type 'string'.
```
I dislike this test.  You aren't testing what you are suggesting.  The server_dyn_res script does return a string.  It returns "100" (not the number, but the string "100").  Nothing is really going wrong here.  The job is not running because your strings do not match.  I say delete this test.

latha-subramanian replied:
Yes. Realized late that couple of these tests are not adding any value as the script is always adding a string type. I am deleting those tests.

bhroam commented on test/tests/functional/pbs_svr_dyn_res.py:
```diff
@@ -480,3 +480,237 @@ def test_res_size_runtime(self):
+        """
+        Test invalid values returned from server_dyn_resource
+        script for resource type 'string'
+        Script returns a 'float' instead of type 'string'.
```
Same as above, your job is not running because the strings do not match.

bhroam commented on test/tests/functional/pbs_svr_dyn_res.py:
```diff
@@ -480,3 +480,237 @@ def test_res_size_runtime(self):
+        self.setup_dyn_res(resname, restype, resval)
+
+        # Submit job
+        a = {'Resource_List.foo': '"red,blue"'}
```
String arrays are a one to N match.  You don't request "red,blue", you request "red" or "blue".  You can't request multiple values of a string array.  Your string array has one value in it ("red blue").  You are requesting the string "red,blue" and it does not match the one value "red blue".  You aren't really testing anything special here.

bhroam commented on test/tests/functional/pbs_svr_dyn_res.py:
```diff
@@ -480,3 +480,237 @@ def test_res_size_runtime(self):
+        self.setup_dyn_res(resname, restype, resval)
+
+        # Submit job
+        a = {'Resource_List.foo': '"1,2,3"'}
```
You are requesting a string "1,2,3", not three values of a string array.  If you were to request 1 or 2 or 3, your job would run.  The server_dyn_res script is not returning ints.  It's returning three values for the string array ("1", "2", and "3").

bhroam commented on test/tests/functional/pbs_svr_dyn_res.py:
```diff
@@ -480,3 +480,237 @@ def test_res_size_runtime(self):
+        job_comment = "Can Never Run: Insufficient amount of server resource:"
+        job_comment += " foo (R: 1.2 A: 0 T: 0)"
+        a = {'job_state': 'Q', 'comment': job_comment}
+        self.server.expect(JOB, a, id=jid, attrop=PTL_AND)
```
Once again, you can do a log match for the error message if you want.

bhroam commented on test/tests/functional/pbs_svr_dyn_res.py:
```diff
@@ -480,3 +480,237 @@ def test_res_size_runtime(self):
+        job_comment = "Can Never Run: Insufficient amount of server resource:"
+        job_comment += " foo (True != False)"
+        a = {'job_state': 'Q', 'comment': job_comment}
+        self.server.expect(JOB, a, id=jid, attrop=PTL_AND)
```
Once again you can search for the error message if you want.

latha-subramanian commented:
@bhroam ,@dinesh, Please take a look at the latest changes. As pointed out by you I have removed the tests that don't seem to add any value and incorporated other review comments.

[pbs_svr_dyn_res.log](https://github.com/PBSPro/pbspro/files/1926148/pbs_svr_dyn_res.log)

bhroam commented on test/tests/functional/pbs_svr_dyn_res.py:
```diff
@@ -502,148 +496,20 @@ def test_res_size_invalid_input(self):
-        a = {'Resource_List.foo': '"1,2,3"'}
-        j = Job(TEST_USER, attrs=a)
-        jid = self.server.submit(j)
+        # Also check that "Script %s returned bad output"
```
Change the %s to <scr> or <script> (and for other occurrences in the test suite)

latha-subramanian replied:
Fixed it.
[pbs_svr_dyn_res.log](https://github.com/PBSPro/pbspro/files/1926184/pbs_svr_dyn_res.log)

bhroam commented on test/tests/functional/pbs_svr_dyn_res.py:
```diff
@@ -112,7 +109,7 @@ def test_invalid_script_out(self):
                                  existence=False, max_attempts=10)
 
-        # Also check that "Script %s returned bad output"
+        # Also check that "<script> returned bad output"
```
How about: Also check that "<script> returned bad output" is in the logs.

latha-subramanian commented:
Thanks @bhroam , @dinesh1306 . I have squashed and rebased.
@PBSPro/pbspro-maintainers, PR is ready for merge. Please take it forward. Thanks!

bhroam commented:
@latha-subramanian the PR is now behind.  Please rebase again.

latha-subramanian commented:
@bhroam I have rebased again. Please take it forward.

bhroam commented:
I merged in the PR.
